sql/hive/src/main/java/org/apache/spark/sql/hive/mapred/CombineSplitRecordReader.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.hive.mapred;
+
+import java.io.IOException;
+
+import org.apache.hadoop.classification.InterfaceAudience;
+import org.apache.hadoop.classification.InterfaceStability;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.mapred.InputFormat;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hadoop.mapred.RecordReader;
+import org.apache.hadoop.mapred.Reporter;
+
+/**
+ * A generic RecordReader that can hand out different recordReaders
+ * for each split in a {@link org.apache.spark.sql.hive.mapred.CombineSplit}.
+ */
+@InterfaceAudience.Public
+@InterfaceStability.Stable
+public class CombineSplitRecordReader<K, V> implements RecordReader<K, V> {
+  protected CombineSplit split;
+  protected JobConf jc;
+  protected FileSystem fs;
+
+  protected int idx;
+  protected long progress;
+  protected RecordReader<K, V> curReader;
+
+  @Override
+  public boolean next(K key, V value) throws IOException {
+    while ((curReader == null) || !curReader.next(key, value)) {
+      if (!initNextRecordReader()) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public K createKey() {
+    return curReader.createKey();
+  }
+
+  public V createValue() {
+    return curReader.createValue();
+  }
+
+  /**
+   * return the amount of data processed
+   */
+  public long getPos() throws IOException {
+    return progress;
+  }
+
+  public void close() throws IOException {
+    if (curReader != null) {
+      curReader.close();
+      curReader = null;
+    }
+  }
+
+  /**
+   * return progress based on the amount of data processed so far.
+   */
+  public float getProgress() throws IOException {
+    return Math.min(1.0f,  progress/(float)(split.getLength()));
```

Comment:
I think the current one can give a more fine-grained metric. i.e. combinedSplit(split1-10M, split2-10M, split3-80M), let's say we've consumed split1 and split2 if we only calc by index then the progress is (2/3), but it would be (20%) otherwise.  Maybe I need to rename `split` -> `combinedSplit` to make name more readable.